File: it-support-engineer/log_analysis.py
#!/user/bin/python 3.8.9
import re
import json
import requests

# 以只读方式读取日志文件
with open('interview_data_set', 'r') as f:
    lines = f.readlines()
# 准备接收目标日志内容的数据结构，列表
valid_lines = []
# 遍历日志，把匹配规则的行添加到valid_lines中，准备后续分析。
for i in range(len(lines)):
    a = [item for item in lines[i].split(' ', 6)]
    try:
        if re.split('\(|\[|\]|:', a[5])[2].isdigit():
            # 正则分隔字符串以判断特征：第6个元素不符合形如'(xxx[000])'的行都不是分析的目标
            valid_lines.append(a)
        else:
            # 不符合特征的第一种情况：能正常正则分割，但不匹配规则，对应位置不是纯数字。
            print('line %s is not error log' % i)
    except Exception as e:
        # 不符合特征的第二种情况：正则分割分不出来第三个元素，out of range。
        print(i, e)
# 准备接收整理完格式日志的列表
formatted_lines = []
# 遍历日志，整理成目标格式。这里都是单条记录，不计数，最后统一计算。
for item in valid_lines:
    deviceName = item[3]
    processId = re.split('\(|\[|\]|:', item[5])[2]
    processName = re.split('\(|\[|\]|:', item[5])[1]
    description = item[6].replace('\n', '')
    """
    拿到时间格式字符串，按':'分割，取第一个元素确定时间段。
    '-'前面的部分3、4位补0，
    '-'后面的部分先变成整数，加1，再变成字符串，不足两位的向前补0，再把3、4位补上0    
    """
    timeWindow = item[2].split(':')[0] + '00-' + str(int(item[2].split(':')[0]) + 1).zfill(2) + '00'
    # 构造json格式数据，暂时没有processID和计数的key，那两部分需要后续处理。
    temp = [{'deviceName': deviceName, 'processName': processName, 'description': description,
             'timeWindow': timeWindow}, processId]
    formatted_lines.append(temp)

# 准备事件列表：对除了processID以外，完全一样的消息去重
event_list = []
for item in formatted_lines:
    if item not in event_list:
        event_list.append(item[0])
# 准备计算列表：把同时段，相同事件的processID以列表形式添加到字典里，每个事件是一个列表元素。
calculation_list = []
for i in event_list:
    processId_list = []
    for j in formatted_lines:
        if i == j[0]:
            processId_list.append(j[1])
    i['processID'] = processId_list
    calculation_list.append(i)
# 准备jason数据列表
json_list = []
for item in calculation_list:
    item['numberOfOccurrence'] = len(item['processID'])
    # processID列表的长度就是同时段出现相同事件的次数，把它也添加到json数据结构中。
    json_list.append(item)

json_str = json.dumps(json_list, ensure_ascii=False)
with open('output.txt','w') as f1:
    f1.write(json_str)
"""
最后一步一直报如下证书错误，代码被注释掉了。输出结果请参考output.txt。
requests.exceptions.SSLError: HTTPSConnectionPool(host='foo.com', port=443): 
Max retries exceeded with url: /bar (Caused by SSLError(SSLCertVerificationError(1, 
'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)')))
"""
if json_str:
    url = "https://foo.com/bar"
    headers = {'Content-Type': 'application/json; charset=UTF-8'}
    res = requests.request("post", url, json=json_str, headers=headers)
    print(res.status_code)
    print(res.text)

File: it-support-engineer/logAnalysis.go
package main

import (
	"bufio"
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

type LogFormat struct {
	DeviceName         string
	ProcessId          string
	ProcessName        string
	Description        string
	TimeWindow         string
	NumberOfOccurrence int
}

var logStats map[string]*LogFormat
// 时间窗定义
var time_windows = map[string]string{
	"00": "0000-0100",
	"01": "0100-0200",
	"02": "0200-0300",
	"03": "0300-0400",
	"04": "0400-0500",
	"05": "0500-0600",
	"06": "0600-0700",
	"07": "0700-0800",
	"08": "0800-0900",
	"09": "0900-1000",
	"10": "1000-1100",
	"11": "1100-1200",
	"12": "1200-1300",
	"13": "1300-1400",
	"14": "1400-1500",
	"15": "1500-1600",
	"16": "1600-1700",
	"17": "1700-1800",
	"18": "1800-1900",
	"19": "1900-2000",
	"20": "2000-2100",
	"21": "2100-2200",
	"22": "2200-2300",
	"23": "2300-0000",
}

func main() {
	var logfile string

	logStats = make(map[string]*LogFormat, 0)
	flag.StringVar(&logfile, "f", "", "日志文件名称")
	flag.Parse()

	fmt.Println(logfile)
	if logfile == "" {
		log.Fatal(errors.New("未输入参数"))
	}
	// 读取日志
	f, err := os.Open(logfile)
	if err != nil {
		fmt.Println(err.Error())
	}
	log.Printf("开始分析日志文件%s", logfile)
	//建立缓冲区，把文件内容放到缓冲区中
	buf := bufio.NewReader(f)
	for {
		//遇到\n结束读取
		b, errR := buf.ReadBytes('\n')
		if errR != nil {
			if errR == io.EOF {
				break
			}
			fmt.Println(errR.Error())
		}

		//解析正则表达式，如果成功返回解释器
		reg := regexp.MustCompile(`(.+\s\d+:\d+:\d+?)\s(.+?)\s(.+?):\s(.+)`)
		if reg == nil {
			fmt.Println("regexp err")
			return
		}

		result := reg.FindAllStringSubmatch(string(b), -1)
		//range遍历数组
		for _, line := range result {
			if line[0] == " " {
				return
			}
			analyLog(line)
		}
	}
	if logStats != nil {
		logTimeStats := make([]LogFormat, 0)
		for _, v := range logStats {
			logTimeStats = append(logTimeStats, *v)
		}
		data, _ := json.Marshal(logTimeStats)
		dataPost(string(data))
	}
	log.Println("分析日志完成")
}

func analyLog(line []string) {
	dateTime := strings.Split(line[1], " ")[2]
	item := &LogFormat{}
	item.DeviceName = line[2]
	item.TimeWindow = time_windows[strings.Split(dateTime, ":")[1]]
	process := strings.Split(line[3], " ")
	for k, v := range process {
		if k == 0 { // 假设需要获取os.Args[k], k = 1
			item.ProcessId = v
		}
		if k == 1 { // 假设需要获取os.Args[k], k = 1
			item.ProcessName = v
		}
	}
	item.Description = line[4]

	idx := item.TimeWindow + "|" + item.DeviceName + "|" + item.ProcessId + "|" + item.ProcessName + "|" + item.Description
	idxmd5 := md5V(idx)
	if _, ok := logStats[idxmd5]; ok {
		logStats[idxmd5].NumberOfOccurrence = logStats[idxmd5].NumberOfOccurrence + 1
	} else {
		item.NumberOfOccurrence = 1
		logStats[idxmd5] = item
	}
}

// post数据
func dataPost(data string) {
	fmt.Println("post data=", string(data))

	resp := Post("https://foo.com/bar", data, "application/json")
	//resp := Post("http://127.0.0.1:5566/", data, "application/json")
	log.Printf("分析数据成功:%s", resp)
}

// post请求封装
func Post(url string, data interface{}, contentType string) (content string) {
	jsonStr, _ := json.Marshal(data)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
	req.Header.Add("content-type", contentType)
	if err != nil {
		panic(err)
	}
	defer req.Body.Close()

	client := &http.Client{Timeout: 5 * time.Second}
	resp, error := client.Do(req)
	if error != nil {
		panic(error)
	}
	defer resp.Body.Close()

	result, _ := ioutil.ReadAll(resp.Body)
	content = string(result)
	return
}


func md5V(str string) string  {
	h := md5.New()
	h.Write([]byte(str))
	return hex.EncodeToString(h.Sum(nil))
}